plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java

```diff
@@ -0,0 +1,455 @@
+            else {
+                close();
+            }
+            return null;
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return immediateVoidFuture();
+        }
+
+        @Override
+        public synchronized long getRetainedSize()
+        {
+            return INSTANCE_SIZE + (sliceInput == null ? 0 : sliceInput.getRetainedSize());
+        }
+
+        @Override
+        public synchronized boolean isFinished()
+        {
+            return closed;
+        }
+
+        @Override
+        public synchronized void close()
+        {
+            if (closed) {
+                return;
+            }
+            closed = true;
+            if (sliceInput != null) {
+                sliceInput.close();
+                sliceInput = null;
+            }
+        }
+    }
+
+    @NotThreadSafe
+    private static class AzureExchangeStorageWriter
+            implements ExchangeStorageWriter
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(AzureExchangeStorageWriter.class).instanceSize();
+
+        private final BlockBlobAsyncClient blockBlobAsyncClient;
+        private final int blockSize;
+
+        private ListenableFuture<Void> directUploadFuture;
+        private final List<ListenableFuture<Void>> multiPartUploadFutures = new ArrayList<>();
+        private final List<String> blockIds = new ArrayList<>();
+        private volatile boolean closed;
+
+        public AzureExchangeStorageWriter(
+                BlockBlobAsyncClient blockBlobAsyncClient,
+                int blockSize)
+        {
+            this.blockBlobAsyncClient = requireNonNull(blockBlobAsyncClient, "blockBlobAsyncClient is null");
+            this.blockSize = blockSize;
+        }
+
+        @Override
+        public ListenableFuture<Void> write(Slice slice)
+        {
+            checkState(directUploadFuture == null, "Direct upload already started");
+            if (closed) {
+                // Ignore writes after writer is closed
+                return immediateVoidFuture();
+            }
+
+            // Skip multipart upload if there would only be one part
+            if (slice.length() < blockSize && multiPartUploadFutures.isEmpty()) {
+                directUploadFuture = asVoid(toListenableFuture(blockBlobAsyncClient.upload(Flux.just(slice.toByteBuffer()), slice.length()).toFuture()));
+                return directUploadFuture;
+            }
+
+            String blockId = Base64.getEncoder().encodeToString(UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8));
+            ListenableFuture<Void> uploadFuture = toListenableFuture(blockBlobAsyncClient.stageBlock(blockId, Flux.just(slice.toByteBuffer()), slice.length()).toFuture());
+            multiPartUploadFutures.add(uploadFuture);
+            blockIds.add(blockId);
+            return uploadFuture;
+        }
+
+        @Override
+        public ListenableFuture<Void> finish()
```

Comment:
would that make sense to mark Writer as `finishing` so we throw an error if `finish` is called twice (or we can return previous future for second call).

(can be a followup - if to be done, same applies for s3 implementation)

Reply:
Actually in `ExchangeSink`'s protocol, for finish() `This method is guaranteed not be called more than once.`.